### PR TITLE
Pwd min length

### PR DESCRIPTION
**Pull request by vertexian**
So I think I've got things in a mergeable state. The main changes are backend: adding a password complexity validation class (currently just requires length ≥ 5), and then invoking it in three places:
- Initial sign-up
- Profile edit
- Forgotten password reset
I also had to pass through a validation errors param and style it on the front-end. I took the liberty of adding an alert icon (CC licensed from the Noun Project; we might want to take more icons from that set).

A few implementation notes:
- Validation is implemented by throwing and catching a MemberError, which seems like the Pythonic approach (I've used other languages that are stingier with exceptions). Since I had that mechanism I went ahead and used it for is-superuser validation.
- I added a lookup_member_key() method to reduce code repetition. That's the biggest risk area for regressions, I think.
- Finally, I tried to get localization working by writing my error string as a _("") string. However, in my testing I wasn't able to get it to substitute, and I haven't figured out how to attach a debugger yet. (Using PyCharm at the moment; may try Komodo as well.)

Let me know what you think!

### CODE REVIEW

**Comment by vertexian** [comment]
I figured out how to do the localization of strings outside of templates, and also found a bug with the alert icon displaying when it shouldn't. Will make another commit for these later today.

**Comment by vertexian** [comment]
Solved the i18n problem (we now set up i18n at request dispatch time and overload later if there's a `locale=` param) and addressed the alert icon display bugs. Should be ready for review.

**Comment by vertexian** [comment]
My fiancée just took over my workstation to do some git magic and remove the unwanted changesets from the branch. Should just be mine in there now.

**Comment by aaronoppenheimer** [comment]
You should marry that person. I'm traveling but will take a look Thursday
when I get home.

A

File: member.py [hunk 1]
```diff
@@ -957,7 +936,7 @@ def post(self):
             the_search_str=self.request.get('s', None)
             self._make_page(the_user=self.user, the_page=the_page, the_search=the_search_str)
         else:
-            return;
+            self.abort_nonadmin()
```
Comment (aaronoppenheimer) [inline member.py]: should this be abort_non_admin()?

Comment (aaronoppenheimer) [inline member.py]: Actually, we have a @superuser_required decorator - should probably do this that way instead. Weirdly, the only place it's in use is in the "forums" module, which is half-baked and not exposed. But it's a neater way to go.

File: member.py [hunk 2]
```diff
@@ -1045,10 +1021,11 @@ def get(self):
...
-        
-        if (the_user != the_member_key):
+
+        if not member_is_superuser(self.user):
+            self.abort_non_admin()
+
+        if (self.user != the_member_key):
             forget_member_from_key(the_member_key)
         else:
             print 'cannot delete yourself, people'
```
Comment (aaronoppenheimer) [inline member.py]: Should probably throw an exception. The only way to get here is to manually paste your key into the url - someone doing that is a troublemaker.

File: templates/signup.html [hunk 3]
```diff
@@ -22,7 +22,7 @@
         {% if failed %}
             <div class="row">
                 <div class="col-md-5 col-md-offset-2">
-                    <div class="text-error">{{ error }}</div>
+                    <div class="validation-error">{{ error }}</div>
```
Comment (aaronoppenheimer) [inline templates/signup.html]: Is this actually styled someplace?

File: requestmodel.py [hunk 4]
```diff
@@ -168,15 +160,24 @@ def display_message(self, message):
...
-				# Dispatch the request.
-				webapp2.RequestHandler.dispatch(self)
-		finally:
-				# Save all sessions.
-				self.session_store.save_sessions(self.response)
+        # Get a session store for this request.
+        self.session_store = sessions.get_store(request=self.request)
+        # Set locale on each request right away, to make it available to both
+        # templates and backend Python code
+        if self.request.get("locale"):
```
Comment (aaronoppenheimer) [inline requestmodel.py]: Good idea to do it here. Getting the translation stuff to work is a pain if you step outside the templating system. I'll have to track down the other non-rendering places I've had to mess with it - in the email and caldav modules - maybe this cleans that up, too.

File: css/gigo.css [hunk 5]
```diff
@@ -112,6 +112,16 @@ code {
     overflow:visible;
 }
 
+.validation-error {
+    margin-bottom: 20px;
+}
+
+.validation-error::before {
+    content: url('./images/alert.png');
```
Comment (vertexian) [inline css/gigo.css]: Done. Good to know we already have a standard icon set and I just missed it.

The rest of the review changes will have to wait a bit, as I am indeed marrying she-of-the-mad-git-skillz, this very day in fact! 💒🎉

**Comment by aaronoppenheimer** [comment]
WOW! Congratulations!

**Comment by vertexian** [comment]
I didn't see any locale code in the caldav modules. I removed locale-setting from the email code where the email originated from a web request.

I didn't test every instance of emailing, but the ones I did (and the tests I did on @superuser_required) were successful.

**Comment by aaronoppenheimer** [comment]
Ew. Know what I just remembered? We already have a password validation system in place. If you look in "member_edit.html" at the bottom are some rules for jquery.validate - if we just add a rule for the field "member_password1" using "minlength: 5" we're good to go; can add a translatable error message, too.

I'd prefer doing it this way rather than making a round trip to the server - which means you did a bunch of work here you didn't have to do. I'm sorry!

But - all the refactoring and introduction of exceptions is important. What's the way to keep that but back out the length validation? And is the other pull-request completely separate from this one?

**Comment by vertexian** [comment]
On 2017-05-08 17:03, aaronoppenheimer wrote:

> I'd prefer doing it this way rather than making a round trip to the server - which means you did a bunch of work here you didn't have to do. I'm sorry!

I can back out the validation, but I'm not sure I agree that having the
server check is a bad thing. I agree that if it's trivial in the
front-end then we can do that, but I don't think it hurts to always have
a server check of any client-side validation. That we don't accidentally
bypass it. For instance, it's not just member_edit that sets the
password, it's also the initial signup and password reset pages. 

Can I add the jQuery check to the existing code (using the same error
message) and go forward with that? 

Also, I can make a list of new strings for translators. When I re-ran
the translation build script, it picked up some other, unrelated strings
that haven't been translated yet. 

Mike

**Comment by aaronoppenheimer** [comment]
Yes, you're right - we should validate on the server side anyway. Agree that you can just add the validation rules to the validator on member_edit, signup, and resetpassword pages - they all have the same validator.

Re: the translation strings - yeah, there are a few old ones in there that I didn't bother with. After this merge is through, I'll take care of it. Thanks, Mike!

**Comment by aaronoppenheimer** [comment]
Done!

**Pull request merged**